src/components/Questionnaire/ValueSetSelect.tsx

```diff
@@ -67,12 +83,127 @@ export default function ValueSetSelect({
+      pathParams: { slug: system },
+    }),
+    onSuccess: () => {
+      queryClient.invalidateQueries({
+        queryKey: ["valueset", system, "favourites"],
+      });
+    },
+  });
+
+  const clearFavouritesMutation = useMutation({
+    mutationFn: mutate(valuesetRoutes.clearFavourites, {
+      pathParams: { slug: system },
+    }),
+    onSuccess: () => {
+      queryClient.invalidateQueries({
+        queryKey: ["valueset", system, "favourites"],
+      });
+      setIsClearingFavourites(false);
+    },
+  });
+
+  const recentsQuery = useQuery({
+    queryKey: ["valueset", system, "recents"],
+    queryFn: query(valuesetRoutes.recentViews, {
+      pathParams: { slug: system },
+    }),
+  });
+
+  const addRecentMutation = useMutation({
+    mutationFn: mutate(valuesetRoutes.addRecentView, {
+      pathParams: { slug: system },
+    }),
+    onSuccess: () => {
+      queryClient.invalidateQueries({
+        queryKey: ["valueset", system, "recents"],
+      });
+    },
+  });
+
+  // Combine recents and search results, but only show each result once
+  const resultsWithRecents = [
+    ...(recentsQuery.data?.filter((recent) =>
+      recent.display?.toLowerCase().includes(search.toLowerCase()),
+    ) || []),
+    ...(searchQuery.data?.results?.filter(
+      (r) => !recentsQuery.data?.find((recent) => recent.code === r.code),
+    ) || []),
+  ];
+
+  // Filter favourites based on search
+  const favourites = favouritesQuery.data?.filter((favourite) =>
+    favourite.display?.toLowerCase().includes(search.toLowerCase()),
+  );
+
   useEffect(() => {
     if (controlledOpen || internalOpen) {
       setSearch("");
     }
   }, [controlledOpen, internalOpen]);
 
+  const Item = ({ option }: { option: Code }) => (
+    <CommandItem
+      key={option.code}
+      value={option.code}
+      onSelect={() => {
+        onSelect({
+          code: option.code,
+          display: option.display || "",
+          system: option.system || "",
+        });
+        setInternalOpen(false);
+        addRecentMutation.mutate(option);
+      }}
+      className="cursor-pointer"
+    >
+      <div className="flex items-center justify-between w-full gap-4">
+        <span>{option.display}</span>
+
+        <button
+          type="button"
+          onClick={(e) => {
+            e.preventDefault();
+            e.stopPropagation();
+            favouritesQuery.data?.find(
+              (favourite) => favourite.code === option.code,
+            )
+              ? removeFavouriteMutation.mutate(option)
+              : addFavouriteMutation.mutate(option);
+          }}
+          className="hover:text-primary-500 transition-all text-secondary-900"
+        >
+          {favouritesQuery.data?.find(
+            (favourite) => favourite.code === option.code,
+          ) ? (
+            <StarFilledIcon />
+          ) : (
+            <StarIcon className="" />
+          )}
+        </button>
+      </div>
+    </CommandItem>
+  );
```

Comment:
avoid defining components inside a component. react will not memoize this correctly since the reference of `Item` keeps changing on every re-render.